### PR TITLE
Al lot of stuff

### PR DESCRIPTION
**Pull request by AndreaLanfranchi**
This is a huge rework.

## Relevant topics
1. **CLI Interface** : `CLI11` is used only to validate arguments while all texts have been rewritten. Help is now split in sections. Base help appear like this
```
ethminer 0.17.0-alpha.1+commit.f60923d8
Build: windows/release/msvc

Ethminer - GPU ethash miner
minimal usage : ethminer [DEVICES_TYPE] [OPTIONS] -P... [-P...]

Devices type options :

    By default ethminer will try to use all devices types
    it can detect. Optionally you can limit this behavior
    setting either of the following options
    -G,--opencl         Mine/Benchmark using OpenCL only
    -U,--cuda           Mine/Benchmark using CUDA only

Connection options :

    -P,--pool           Stratum pool or http (getWork) connection as URL
                        scheme://[user[.workername][:password]@]hostname:port[/...]
                        For an explication and some samples about
                        how to fill in this value please use
                        ethminer --help-ext con

Common Options :

    -h,--help           Displays this help text and exits
    -H,--help-ext       TEXT {'con','test',cl,cu,api,'misc','env'}
                        Display help text about one of these contexts:
                        'con'  Connections and their definitions
                        'test' Benchmark and simulation options
                        'cl'   Extended OpenCL options
                        'cu'   Extended CUDA options
                        'api'  API and Http monitoring interface
                        'misc' Other miscellaneous options
                        'env'  Using environment variables
    -V,--version        Show program version and exits
```
2. **CLI flow** : all validation process throws exceptions to be handled in main(). Main thread is also put in a non-busy wait state by the means of a condition variable. In presence of NO cli arguments the thread exits with help message (instead of producing unexpected behaviors). Presence of pool definition is not checked if the run mode does not require that (eg. --list-devices)
3. Solution submission redesigned to work async
4. Pool rotation : avoid infinite loop when only one pool specified
5. **Stratum** : avoided possible double reads and read all transmissions as whole. Only last job of each transmission is dispatched
6. **GetWork** : client reworked to run async and removed it's own thread. Removed dependency for json-rpc-cpp.lib
7. **Mixed mode**: all usable devices are abstracted and miners are labelled with sequential indexes (no more cuda-0 and cl-0 miners). Report of device PCI Id. Mixed mode now works as excpected and by default is on. -U and -G are used if user wants to **limit** devices in use
8. **DAG Load mode 2 (single)** dropped

### CODE REVIEW

File: libpoolprotocols/stratum/EthStratumClient.cpp [hunk 1]
```diff
@@ -308,7 +308,7 @@ void EthStratumClient::disconnect_finalize()
 {
     if (m_conn->SecLevel() != SecureLevel::NONE)
     {
-        if (m_securesocket->lowest_layer().is_open())
+        if (m_securesocket && m_securesocket->lowest_layer().is_open())
         {
             // Manage error code if layer is already shut down
             boost::system::error_code ec;
```
Comment (jean-m-cyr) [inline libpoolprotocols/stratum/EthStratumClient.cpp]: Deleting lines 319-323 would reduce duplication of common code sequence.

**Comment by jean-m-cyr** [comment]
CL_DEVICE_COMPUTE_CAPABILITY_MAJOR_NV and CL_DEVICE_COMPUTE_CAPABILITY_MINOR_NV are not supported in opencl 1.0.
You can do something like:
```
#ifdef CL_DEVICE_COMPUTE_CAPABILITY_MAJOR_NV
    device.getInfo<CL_DEVICE_COMPUTE_CAPABILITY_MAJOR_NV>();
#else 
   ...
#endif 
```

**Comment by AndreaLanfranchi** [comment]
@jean-m-cyr the only case ethminer can run on opencl 1.0 is with clover platform. Should we keep it ?

**Comment by jean-m-cyr** [comment]
Not working here!!!

```
jcyr@miner0:~$ /usr/local/bin/ethminer -U -R --display-interval 30 --noeval -v 2 $@ \
>         --api-port 40000 \
>         -P stratum+tls12://0xB73xxx101.$HOSTNAME@us1.ethermine.org:5555 \
>         -P stratum+tls12://0x839xxx18F.$HOSTNAME@us1-etc.ethermine.org:5555


ethminer 0.18.0-alpha.0+commit.1c76c9a6
Build: linux/release/gnu

 i 17:02:22 ethminer Configured pool us1.ethermine.org:5555
 i 17:02:22 ethminer Configured pool us1-etc.ethermine.org:5555
 i 17:02:22 ethminer Api server listening on port 40000.
 i 17:02:22 ethminer Selected pool us1.ethermine.org:5555
 i 17:02:23 ethminer Stratum mode detected: STRATUM
 i 17:02:23 ethminer Subscribed!
 i 17:02:23 ethminer Authorized worker 0xB73Fb0A12Cc203e479Db2A49D68C3076F9d45101
 i 17:02:23 ethminer Established connection to us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:02:23 ethminer Spinning up miners...
cu 17:02:23 cuda-0   Using Pci Id : 04:00.0 GeForce GTX 1060 6GB (Compute 6.1) Memory : 5.93 GB
cu 17:02:23 cuda-1   Using Pci Id : 05:00.0 GeForce GTX 1060 3GB (Compute 6.1) Memory : 2.95 GB
cu 17:02:23 cuda-2   Using Pci Id : 06:00.0 GeForce GTX 1060 3GB (Compute 6.1) Memory : 2.95 GB
 i 17:02:23 ethminer Epoch : 0
 i 17:02:23 ethminer Difficulty : 4.00Gh
 i 17:02:23 ethminer Job: #33cde6ed… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
cu 17:02:23 cuda-3   Using Pci Id : 0a:00.0 GeForce GTX 1060 3GB (Compute 6.1) Memory : 2.95 GB
cu 17:02:25 cuda-0   Generating DAG + Light : 1.02 GB
cu 17:02:25 cuda-3   Generating DAG + Light : 1.02 GB
cu 17:02:25 cuda-1   Generating DAG + Light : 1.02 GB
cu 17:02:25 cuda-2   Generating DAG + Light : 1.02 GB
cu 17:02:30 cuda-3   Generated DAG + Light in 6,506 ms. 1.93 GB left.
cu 17:02:30 cuda-0   Generated DAG + Light in 6,509 ms. 4.92 GB left.
cu 17:02:30 cuda-1   Generated DAG + Light in 6,645 ms. 1.93 GB left.
cu 17:02:30 cuda-2   Generated DAG + Light in 6,652 ms. 1.93 GB left.
 i 17:02:41 ethminer Job: #60bfd46c… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:02:44 ethminer Job: #7562e250… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:02:44 ethminer Job: #41c251df… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:02:45 ethminer Job: #a71e3cd3… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:02:48 ethminer Job: #c08d5777… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:02:49 ethminer New API session from 192.168.1.169:37822
 m 17:02:52 ethminer Speed 84.76 Mh/s gpu0 21.98 gpu1 22.19 gpu2 22.19 gpu3 18.40 Time: 00:00
 m 17:02:52 ethminer Solutions A0 gpu0:A0 gpu1:A0 gpu2:A0 gpu3:A0
 i 17:02:58 ethminer Job: #51766338… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:03:06 ethminer Job: #5115cef2… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:03:07 ethminer Job: #fa132c47… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:03:08 ethminer Job: #5455f98f… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:03:13 ethminer Job: #4c85f483… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:03:14 ethminer Job: #720a26ac… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:03:14 ethminer Job: #7e90fb04… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 m 17:03:22 ethminer Speed 84.76 Mh/s gpu0 21.97 gpu1 22.19 gpu2 22.19 gpu3 18.40 Time: 00:01
 m 17:03:22 ethminer Solutions A0 gpu0:A0 gpu1:A0 gpu2:A0 gpu3:A0
 i 17:03:30 ethminer Job: #051a7d36… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 m 17:03:52 ethminer Speed 84.77 Mh/s gpu0 21.97 gpu1 22.20 gpu2 22.19 gpu3 18.40 Time: 00:01
 m 17:03:52 ethminer Solutions A0 gpu0:A0 gpu1:A0 gpu2:A0 gpu3:A0
 i 17:04:15 ethminer Job: #d53b632c… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:04:21 ethminer Job: #aa584a7d… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 m 17:04:22 ethminer Speed 84.76 Mh/s gpu0 21.97 gpu1 22.20 gpu2 22.19 gpu3 18.40 Time: 00:02
 m 17:04:22 ethminer Solutions A0 gpu0:A0 gpu1:A0 gpu2:A0 gpu3:A0
cu 17:04:39 cuda-3   Sol: 0xbd43e455b45d1a4c
 X 17:04:39 ethminer Reject reason :Provided PoW solution is invalid!
 X 17:04:39 ethminer **Rejected  35ms.   us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:04:42 ethminer Job: #9c30dcf5… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:04:42 ethminer Job: #0a8f2adf… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:04:44 ethminer Job: #710e7cf7… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
 i 17:04:47 ethminer Job: #c1cc6130… block 250 us1.ethermine.org:5555 [[2606:4700:90:0:b886:9f1:5c3e:5711]:5555]
```

**Comment by jean-m-cyr** [comment]
@AndreaLanfranchi I've suggested dropping Apple support, at least the xcode (native) environment. The decision is not mine to make...

**Comment by jean-m-cyr** [comment]
Help looks cleaner now. Unfortunately will be harder to maintain since it does not make use of CLI11's built-in help generation capability.

**Comment by AndreaLanfranchi** [comment]
@jean-m-cyr
There is something wrong with Stratum support. Epoch 0 and block 250 do not make any sense. Try stratum1 or leave autodetect

**Comment by jean-m-cyr** [comment]
I thought stratum+tls12 IS auto-detect??? If I use stratum1+tls12 it works correctly.

**Comment by jean-m-cyr** [comment]
I've approved these changes though I expect the stratum+tls12 scheme will eventually addressed. Also some resolution would need to be reached regarding the Apple xcode build. Either the build error needs to be addressed or Apple native support dropped. Either way, fixing this will require re-approval.

**Comment by jean-m-cyr** [comment]
Hope you don't mind the few quick patches. I think it's good enough to go.

**Comment by StefanOberhumer** [comment]
Just as info (as someone else may observe that):


Recognized that I get now
`cu 09:04:58 cuda-0   Generated DAG + Light in 20022 ms. 3.13 GB left.`
instead of previous
`cu 09:22:36 cuda-0   Generated DAG for GPU0 in: 14019 ms.`

So I searched where the 6 seconds get lost.
The time is "needed" as now the whole DAG generation is measured (including resetting device, allocating memory, copying lightcache to card and generating the "plain DAG").
Previous version just measured the "plain DAG" generation on the card ( ethash_generate_dag() ) which still needs the same time in my environment (~ about 14000ms)

**Comment by AndreaLanfranchi** [comment]
@jean-m-cyr 
> I thought stratum+tls12 is auto-detect??? 

No ... stratum+tls12 is not autodetect: this is like "stratum1+tcp but with tls12 encryption"
Only schemes for autodetection are :
- stratum://
- stratums://
- stratumss://

**Comment by AndreaLanfranchi** [comment]
@StefanOberhumer 

>The time is "needed" as now the whole DAG generation is measured (including resetting device, allocating memory, copying lightcache to card and generating the "plain DAG").

You're right ... previous version did not take into account the whole time needed to initialize a new epoch. This instead computes the whole time.

But there is another tiny difference: in previous version the `init()` performed some operations which were not related to epoch rather to setting device for thread. I know split the initial device setting into `initDevice()` which is performed only once and all the rest needed for epoch initialization is in `initEpoch_internal()`

**Pull request merged**